### DISCUSSION TITLE
How to reduce time in multiprediction in multi thread?

### DISCUSSION THREAD

**Discussion by jihunddok** [ultralytics/ultralytics#9097] [Q&A]
Im gonna make multi model server.
models are single class segmentation model.(14 models)
so, What I`m going to do is 
1. input 1 image.
2. predict in each models.
3. sum and post process to each predicted results.
in this case, I have to use multi thread prediction.
so I make this system by multi thread.
but prediction speed is so slow.
how to enhance prediction speed in multi process safly?
this part of my code.

``` python

def loadModels(self):
    self.model0 = YOLO("./models/modelA.pt")
    self.model1 = YOLO("./models/modelB.pt")
    self.model2 = YOLO("./models/modelC.pt" )
    self.model3 = YOLO("./models/modelD.pt" )
    self.model4 = YOLO("./models/modelE.pt" )
    self.model5 = YOLO("./models/modelF.pt")\
        ...

def startPredictionThread(self):
    th_model1 = Thread(target=self.singlePrediction,args=(self.model0,self.image_path))
    th_model2 = Thread(target=self.singlePrediction,args=(self.model1,self.image_path))
    th_model3 = Thread(target=self.singlePrediction,args=(self.model2,self.image_path))
    th_model4 = Thread(target=self.singlePrediction,args=(self.model3,self.image_path))
    th_model5 = Thread(target=self.singlePrediction,args=(self.model4,self.image_path))
    th_model6 = Thread(target=self.singlePrediction,args=(self.model5,self.image_path))
    
    th_model1.start()
    ...
    th_model6.join()

def singlePrediction(self,model,image):
        pre_img = Image.open(image)
        single_model = model
        results = single_model.predict(pre_img,save=False)
        for res in results:
            content = res.tojson()
            json_content = json.loads(content)
            self.sum_json(content=json_content)
```

**Comment by pderrenger** [comment]
Hey there! It sounds like you're working on a pretty sophisticated multi-model server setup 👏. When it comes to enhancing prediction speed while using multiple threads, here are a few suggestions that might help:

1. **Model Optimization**: If you haven't already, consider optimizing your models before deploying them. This can be done through quantization or pruning, which can significantly reduce inference time while maintaining accuracy. For instance, exporting your models to ONNX and then optimizing them with tools like ONNX Runtime can be beneficial.

2. **Thread Management**: Python has a Global Interpreter Lock (GIL) that can make multi-threading challenging for CPU-bound tasks. If your server's CPU is the bottleneck, using multiprocessing instead of multi-threading could yield better results. This involves changing `Thread` to `Process` from Python's multiprocessing module. Remember, this approach might increase memory usage since each process has its own memory space.

3. **Batch Inference**: If possible, modify your prediction function to perform batch inference instead of predicting images one by one. Most deep learning models, including those in YOLO, are capable of handling multiple inputs at once, which can be much faster than sequential predictions.

4. **GPU Utilization**: Ensure that you're leveraging GPU acceleration if your server's hardware supports it. Running predictions on a GPU can drastically reduce inference time compared to CPU. This might involve setting the device parameter in your YOLO model loading phase to 'cuda' or whatever is applicable to your setup.

5. **Model Serving Frameworks**: Consider using a model serving framework like TensorFlow Serving or NVIDIA Triton Inference Server. These frameworks are designed to handle multiple models and requests efficiently, optimizing resource usage and supporting both batching and multiple threading/processes.

Just as a quick reference, if you decide to try multiprocessing, here's how you might adjust your `startPredictionThread` function:

```python
from multiprocessing import Process

def startPredictionProcess(self):
    th_model1 = Process(target=self.singlePrediction,args=(self.model0,self.image_path))
    th_model2 = Process(target=self.singlePrediction,args=(self.model1,self.image_path))
    # ...
    th_model1.start()
    # ...
    th_model1.join()
```

I hope this helps! Speeding up multi-model inference definitely has its challenges, but with the right approach, you'll see significant improvements. Good luck! 🚀

**Reply by jihunddok** [reply]
I used a multi-model using process as you suggested.
But I found the following error:

load models...
analyze start....
Traceback (most recent call last):
  File "...\multiPredictionMultiprocess_ver04.py", line 207, in <module>
    modelmgr.analyze(image_path="./testimages/testimg.jpg")
  File "...\multiPredictionMultiprocess_ver04.py", line 185, in analyze 
    self.multiPrediction()
  File "...\multiPredictionMultiprocess_ver04.py", line 71, in multiPrediction
    self.model0.start()
  File "\AppData\Local\Programs\Python\Python311\Lib\multiprocessing\process.py", line 121, in start
    self._popen = self._Popen(self)
                  ^^^^^^^^^^^^^^^^^
  File "\AppData\Local\Programs\Python\Python311\Lib\multiprocessing\context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "\AppData\Local\Programs\Python\Python311\Lib\multiprocessing\context.py", line 336, in _Popen
    return Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^
  File "\AppData\Local\Programs\Python\Python311\Lib\multiprocessing\popen_spawn_win32.py", line 95, in __init__
    reduction.dump(process_obj, to_child)
  File "\AppData\Local\Programs\Python\Python311\Lib\multiprocessing\reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
TypeError: cannot pickle '_thread.lock' object

(.yolov8) multicastTCPpython>Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "\AppData\Local\Programs\Python\Python311\Lib\multiprocessing\spawn.py", line 122, in spawn_main
    exitcode = _main(fd, parent_sentinel)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "\AppData\Local\Programs\Python\Python311\Lib\multiprocessing\spawn.py", line 132, in _main
    self = reduction.pickle.load(from_parent)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 EOFError: Ran out of input
 
 
It worked without any problem when I used Thread, but a problem seems to occur when I used Process.

I checked multiprocess problem in predict function. 

because I only tested prediction.

so, is there a problem with the code running for testing?

```python
if __name__=='__main__':
    
    model_mgr = ModelMgr()
    model_mgr.loadModels()
    print("load models")
    start= time.time()
    print("analyze start")
    model_mgr.analyze(image_path="./testimages/testimg.jpg")
    """
    def analyze(self, imagepath):
       self.startPredictionProcess()
    """
    end = time.time()
    print(f"analyze finish : {end - start:.5f} sec")
```

**Reply by pderrenger** [reply]
@jihunddok hey there! 👋 It looks like you're encountering a common issue when moving from threading to multiprocessing with object serialization. When using multiprocessing, Python needs to pickle objects to pass them between processes, and unfortunately, not all objects (like those holding locks or other resources) can be pickled directly.

One workaround is to load your models inside the function called by the `Process` rather than loading them globally or as part of the `__init__` method of your class. This way, the model loading happens within the new process, avoiding the need to pickle complex objects.

Here's a simplified example based on your scenario:

```python
from multiprocessing import Process
import time
from ultralytics import YOLO

def analyze(image_path):
    model = YOLO("./models/modelA.pt")  # Load model here, inside the process
    # Prediction logic here

if __name__=='__main__':
    start= time.time()
    print("analyze start")

    p = Process(target=analyze, args=("./testimages/testimg.jpg",))
    p.start()
    p.join()

    end = time.time()
    print(f"analyze finish : {end - start:.5f} sec")
```

This way, you won't load or attempt to pickle the model object outside of the process. Each process will independently load its required models, thus avoiding the serialization issue you've encountered. Let me know if this helps or if you run into any more issues! 😊
